Make feign.Target.url refreshable

jas34 opened the pull request:
Refer #710

- Created a model class `RefreshableUrl` to wrap url.
- Created `RefreshableUrlFactoryBean` to create instance of `RefreshableUrl` bean with applicable configuration properties.
- Created `RefreshableHardCodedTarget` to serve url wrapped under `RefreshableUrl` whenever needed.
- `url` property added to `FeignClientProperties#FeignClientConfiguration`.
- `FeignClientRegistrar`:
   - created a private method to populate Bean definition for `RefreshableUrl` bean.
   - invoked private method during Feign bean definition registration phase.
 - `FeignClientFactoryBean`:
    - updated to supply `RefreshableHardCodedTarget` target instead of `HardCodedTarget` if URL is defined in the property `spring.cloud.openfeign.client.config.{client-name}.url` and `spring.cloud.openfeign.client.refresh-enabled=true`.
- added test cases.

Fixes gh-710.

OlgaMaciaszek reviewed (changes requested):
Thanks for submitting the PR @jas34. I have added some comments - 1 major about the logic and the rest cosmetic about code style and documentation. Please address the comments. Also please merge new changes from `main` to resolve the conflicts.

jas34 commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java:
```diff
@@ -424,7 +424,7 @@ <T> T getTarget() {
 			}
 			url += cleanPath();
-			return (T) loadBalance(builder, context, new HardCodedTarget<>(type, name, url));
+			return (T) loadBalance(builder, context, (HardCodedTarget<T>) resolveTarget(context, contextId));
```
1. Test case for the same scenario is available test case [withoutConfigUrlFromProperties](https://github.com/jas34/spring-cloud-openfeign/blob/0fdaaa25b1ac1b9a5f42b58f91a3e5ad697fb6c3/spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableUrlTest.java#L63). I have a plan to create the similar regression test case when refresh is disabled.

2. Ok. I will add relevant test case as well.

3.3. 'host is provided in the URL' do you mean URL is provided like this `@FeignClient(name="testClient", url="http://localhost:8081")`? and when you say 'host is provided in the URL' then you mean `@FeignClient(name="testClient")`?
  3.2.a. URL from the property `spring.cloud.openfeign.client.config.{contextId}.url` will be read only when refresh is enabled. Should we add a capbility to read property `spring.cloud.openfeign.client.config.{contextId}.url` for non-refreshable clients as well?

OlgaMaciaszek replied:
Also, I think we should ignore refresh and properties if the users set the URL via annotation. If it's there - they'll get the backwards compatible behaviour. They need to explicitly move it from the annotation to the properties to get the new behaviour. wdyt?

jas34 replied:
- Possible scenarios documented in [spring-cloud-openfeign.adoc](https://github.com/jas34/spring-cloud-openfeign/blob/main/docs/src/main/asciidoc/spring-cloud-openfeign.adoc#supported-ways-to-provide-url-to-a-feign-client)
- Logic implemented as per the documentation.
- Test cases also added to verify the same.

jas34 commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientsRegistrar.java:
```diff
@@ -427,18 +428,35 @@ public void setEnvironment(Environment environment) {
 	 */
 	private void registerOptionsBeanDefinition(BeanDefinitionRegistry registry, String contextId) {
 		if (isClientRefreshEnabled()) {
-			String beanName = Request.Options.class.getCanonicalName() + "-" + contextId;
```
Now calling same private method to register both type of beans.

jas34 commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/RefreshableUrl.java:
```diff
@@ -0,0 +1,37 @@
+
+package org.springframework.cloud.openfeign;
+
+/**
```
Added.

jas34 commented on spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableUrlTest.java:
```diff
@@ -0,0 +1,143 @@
+
+	@Autowired
+	private FeignClientProperties clientProperties;
+
```
Method names changed.

jas34 commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientProperties.java:
```diff
@@ -147,6 +147,8 @@ public static class FeignClientConfiguration {
 
 		private Boolean followRedirects;
 
+		private String url;
```
added.

OlgaMaciaszek commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientProperties.java:
```diff
@@ -147,6 +147,11 @@ public static class FeignClientConfiguration {
 		private Boolean followRedirects;
 
+		/**
+		 * Feign client host URL config definition placeholder.
```
Please convert to sth along the lines of "Allows setting Feign client host URL. This value will only be taken into account if the `url` is not set in the `@FeignClient` annotation." to add more information on how this will work.

jas34 replied:
corrected.

OlgaMaciaszek commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientsRegistrar.java:
```diff
@@ -420,16 +421,17 @@ public void setEnvironment(Environment environment) {
 	/**
-	 * This method is meant to create {@link Request.Options} beans definition with
-	 * refreshScope.
+	 * This method is meant to register beans definition with refreshScope.
```
Please reword from "This method is meant to register" to "This method registers".

jas34 replied:
Done.

OlgaMaciaszek commented on spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableUrlTest.java:
```diff
@@ -0,0 +1,146 @@
+@SpringBootTest
+@TestPropertySource("classpath:feign-refreshable-properties.properties")
+@DirtiesContext
+public class FeignClientWithRefreshableUrlTest {
```
Please rename the class to `FeignClientWithRefreshableUrlTests`.

jas34 replied:
Done.

OlgaMaciaszek commented on spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableUrlTest.java:
```diff
@@ -0,0 +1,146 @@
+@SpringBootTest
+@TestPropertySource("classpath:feign-refreshable-properties.properties")
+@DirtiesContext
+public class FeignClientWithRefreshableUrlTest {
```
Please remove unnecessary `public` modifiers from the class and method signatures.

jas34 replied:
Done

OlgaMaciaszek commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java:
```diff
@@ -467,6 +467,17 @@ private String cleanPath() {
 		return path;
 	}
 
+	private <T> HardCodedTarget<T> resolveTarget(FeignContext context, String contextId) {
```
Ok. Can you add it to `FeignClientProperties ` as well?

jas34 replied:
Added

OlgaMaciaszek reviewed (changes requested):
Thanks for the updates @jas34 . I have resolved some conversations and added some new comments - please address. Also, please get add the logic changes that we've discussed.

galaxy-sea commented:
I like this feature

jas34 commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientsRegistrar.java:
```diff
@@ -427,18 +428,35 @@ public void setEnvironment(Environment environment) {
 	}
 
+	/**
+	 * This method is meant to create {@link RefreshableUrl} beans definition with
```
Done.

jas34 commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/RefreshableUrl.java:
```diff
@@ -0,0 +1,37 @@
+package org.springframework.cloud.openfeign;
+
+/**
+ * This model will just wrap url inside an object so that relevant proxy instance can be
```
Done.

jas34 commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/RefreshableUrl.java:
```diff
@@ -0,0 +1,37 @@
+ */
+public class RefreshableUrl {
+
+	private String url;
```
Done.

jas34 commented:
@OlgaMaciaszek implemented logic for URL as per the above discussion.

OlgaMaciaszek commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java:
```diff
@@ -448,7 +447,7 @@ <T> T getTarget() {
 
 		Targeter targeter = get(context, Targeter.class);
-		return (T) targeter.target(this, builder, context, new HardCodedTarget<>(type, name, url));
+		return (T) targeter.target(this, builder, context, (HardCodedTarget<T>) resolveTarget(context, contextId, url));
```
`(T)` casting is redundant here and can be removed.

jas34 replied:
Removed.

OlgaMaciaszek commented on spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java:
```diff
@@ -467,6 +466,38 @@ private String cleanPath() {
+				return new RefreshableHardCodedTarget<>(type, name, refreshableUrl);
+			}
+		}
+		FeignClientProperties.FeignClientConfiguration config = findConfigByKey(contextId);
```
Not sure why this exception? It's a valid scenario when there's no URL in either properties or annotation and it's being handled in line 414. Please provide your reasoning behind this or fix.

jas34 replied:
In this method, `HardCodedTarget` or `RefreshableHardCodedTarget` needs url so that they can be created as per the applicable constructor. This method accepts url as a parameter or tries to resolve the same from properties if url passed as a param is found to be null/blank. In the case of the unavailability of url, there is a possibility that we return an invalid Target. In order to avoid the same exception has been thrown.

OlgaMaciaszek replied:
Ok, this can't probably happen due to the condition above, but we can leave it in place as an additional safeguard.

OlgaMaciaszek commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,37 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+
+|===
+----
+spring.cloud.openfeign.client.refresh-enabled=true
```
Please add description for this code snippet above.

jas34 replied:
Looks like a typo, added by mistake.

OlgaMaciaszek commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,37 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+|Url is not provided in the `@FeignClient` annotation but provided in configuration properties.
+| `@FeignClient(name="testClient")` and property defined in application.yml as
+`spring.cloud.openfeign.client.config.testClient.url=http://localhost:8081`
+|Url will be resolved from configuration properties without load balancing. if
```
Please hyphenate "load-balancing".

jas34 replied:
Done

OlgaMaciaszek reviewed (changes requested):
Thanks a lot @jas34. This PR is nearly ready for merge. I have added some minor issue comments to the documentation changes, but will also request @Buzzardo to review the docs. I have also added one comment regarding the logic. Please discuss the reasoning behind that or modify it.

jas34 commented:
> Thanks a lot @jas34. This PR is nearly ready for merge. I have added some minor issue comments to the documentation changes, but will also request @Buzzardo to review the docs. I have also added one comment regarding the logic. Please discuss the reasoning behind that or modify it.

All comments fixed.

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -794,7 +795,13 @@ spring.cloud.openfeign.autoconfiguration.jackson.enabled=false
+If Feign client refresh is enabled, each Feign client is created with:
+
+* `feign.Request.Options` as a refresh-scoped bean. This means properties such as `connectTimeout` and `readTimeout` can be refreshed against any Feign client instance.
+* with url wrapped under `org.springframework.cloud.openfeign.RefreshableUrl`. This means url of Feign client if defined
```
Change to:

A URL wrapped under `org.springframework.cloud.openfeign.RefreshableUrl`. This means the URL of the Feign client, if defined

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -794,7 +795,13 @@ spring.cloud.openfeign.autoconfiguration.jackson.enabled=false
+
+* `feign.Request.Options` as a refresh-scoped bean. This means properties such as `connectTimeout` and `readTimeout` can be refreshed against any Feign client instance.
+* with url wrapped under `org.springframework.cloud.openfeign.RefreshableUrl`. This means url of Feign client if defined
+with the property `spring.cloud.openfeign.client.config.{feignName}.url` can be refreshed against any Feign client instance.
```
Change to:

with the `spring.cloud.openfeign.client.config.{feignName}.url` property, can be refreshed against any Feign client instance.

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -794,7 +795,13 @@ spring.cloud.openfeign.autoconfiguration.jackson.enabled=false
+* with url wrapped under `org.springframework.cloud.openfeign.RefreshableUrl`. This means url of Feign client if defined
+with the property `spring.cloud.openfeign.client.config.{feignName}.url` can be refreshed against any Feign client instance.
+
+The above properties can be refreshed through `POST /actuator/refresh`.
```
Change to: 

You can refresh these properties through `POST /actuator/refresh`.

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
 ----
 
+=== Supported Ways To Provide URL To A Feign Client
+URL to a Feign client can be provided in any of the following ways:
```
Change to:

You can provide a URL to a Feign client in any of the following ways:

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+|===
+|Case |Example |Details
+
+|Url is provided in the `@FeignClient` annotation.
```
Change to:

|The URL is provided in the `@FeignClient` annotation.

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+
+|Url is provided in the `@FeignClient` annotation.
+|`@FeignClient(name="testClient", url="http://localhost:8081")`
+|Url will be resolved from url attribute of annotation without load-balancing.
```
Change to:

|The URL is resolved from the `url` attribute of the annotation, without load-balancing.

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+|`@FeignClient(name="testClient", url="http://localhost:8081")`
+|Url will be resolved from url attribute of annotation without load-balancing.
+
+|Url is provided in the `@FeignClient` annotation as well as in
```
Change to:

|The URL is provided in the `@FeignClient` annotation and in the

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+
+|Url is provided in the `@FeignClient` annotation as well as in
+configuration properties.
+|`@FeignClient(name="testClient", url="http://localhost:8081")` and property defined in application.yml as
```
Change to:

|`@FeignClient(name="testClient", url="http://localhost:8081")` and the property defined in `application.yml` as

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+Url provided in the configuration properties remains unused.
+
+|Url is not provided in the `@FeignClient` annotation but provided in configuration properties.
+| `@FeignClient(name="testClient")` and property defined in application.yml as
```
Change to:

| `@FeignClient(name="testClient")` and the property defined in `application.yml` as

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+|Url is not provided in the `@FeignClient` annotation but provided in configuration properties.
+| `@FeignClient(name="testClient")` and property defined in application.yml as
+`spring.cloud.openfeign.client.config.testClient.url=http://localhost:8081`
+|Url will be resolved from configuration properties without load-balancing. if
```
Change to:

|The URL is resolved from configuration properties, without load-balancing. If

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+| `@FeignClient(name="testClient")` and property defined in application.yml as
+`spring.cloud.openfeign.client.config.testClient.url=http://localhost:8081`
+|Url will be resolved from configuration properties without load-balancing. if
+`spring.cloud.openfeign.client.refresh-enabled=true` then url defined in configuration properties can be refreshed as described in <<Spring `@RefreshScope` Support,Spring RefreshScope Support>>.
```
Change to:

spring.cloud.openfeign.client.refresh-enabled=true`, then the URL defined in configuration properties can be refreshed as described in <<Spring `@RefreshScope` Support,Spring RefreshScope Support>>.

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+|Url will be resolved from configuration properties without load-balancing. if
+`spring.cloud.openfeign.client.refresh-enabled=true` then url defined in configuration properties can be refreshed as described in <<Spring `@RefreshScope` Support,Spring RefreshScope Support>>.
+
+|Url is neither provided in the `@FeignClient` annotation nor in configuration properties.
```
Change to:

|The URL is neither provided in the `@FeignClient` annotation nor in configuration properties.

jas34 replied:
Done

Buzzardo commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+
+|Url is neither provided in the `@FeignClient` annotation nor in configuration properties.
+|`@FeignClient(name="testClient")`
+|Url will be resolved from name attribute of annotation with load balancing.
```
Change to:

|The URL is resolved from the name attribute of the annotation, with load balancing.

jas34 replied:
Done

jas34 commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+configuration properties.
+|`@FeignClient(name="testClient", url="http://localhost:8081")` and property defined in application.yml as
+`spring.cloud.openfeign.client.config.testClient.url=http://localhost:8081`
+|Url will be resolved from url attribute of annotation without load-balancing.
```
Done

jas34 commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+|`@FeignClient(name="testClient", url="http://localhost:8081")` and property defined in application.yml as
+`spring.cloud.openfeign.client.config.testClient.url=http://localhost:8081`
+|Url will be resolved from url attribute of annotation without load-balancing.
+Url provided in the configuration properties remains unused.
```
Done

jas34 commented on docs/src/main/asciidoc/spring-cloud-openfeign.adoc:
```diff
@@ -851,6 +858,35 @@ Alternatively, you can use `LoadBalancerFeignRequestTransformer.DEFAULT_ORDER` t
+|Url will be resolved from url attribute of annotation without load-balancing.
+Url provided in the configuration properties remains unused.
+
+|Url is not provided in the `@FeignClient` annotation but provided in configuration properties.
```
Done
